Algorithm.Framework/Portfolio/AccumulativeInsightPortfolioConstructionModel.cs

```diff
@@ -0,0 +1,243 @@
+            _insightCollection.AddRange(insights.Where(ShouldCreateTargetForInsight));
+
+            // Create flatten target for each security that was removed from the universe
+            if (_removedSymbols != null)
+            {
+                var universeDeselectionTargets = _removedSymbols.Select(symbol => new PortfolioTarget(symbol, 0));
+                targets.AddRange(universeDeselectionTargets);
+                _removedSymbols = null;
+            }
+
+            // Get insight that haven't expired of each symbol that is still in the universe
+            var activeInsights = _insightCollection.GetActiveInsights(algorithm.UtcTime);
+
+            // Get the last generated active insight for each symbol
+            var lastActiveInsights = (from insight in activeInsights
+                                      group insight by insight.Symbol into g
+                                      select g.OrderBy(x => x.GeneratedTimeUtc).Last()).ToList();
+
+            var errorSymbols = new HashSet<Symbol>();
+
+            // Determine target percent for the given insights
+            var percents = DetermineTargetPercent(lastActiveInsights);
+
+            foreach (var insight in percents.Keys)
+            {
+                var target = PortfolioTarget.Percent(algorithm, insight.Symbol, percents[insight]);
+                if (target != null)
+                {
+                    targets.Add(target);
+                }
+                else
+                {
+                    errorSymbols.Add(insight.Symbol);
+                }
+            }
+
+            // Get expired insights and create flatten targets for each symbol
+            var expiredInsights = _insightCollection.RemoveExpiredInsights(algorithm.UtcTime);
+
+                        percents = UpdateExpiredInsights(expiredInsights);
+
+            foreach (var insight in percents.Keys)
+            {
+                var target = PortfolioTarget.Percent(algorithm, insight.Symbol, percents[insight]);
+                if (target != null)
+                {
+                    targets.Add(target);
+                }
+                else
+                {
+                    errorSymbols.Add(insight.Symbol);
+                }
+            }
+
+            _nextExpiryTime = _insightCollection.GetNextExpiryTime();
+
+            return targets;
+        }
+
+        /// <summary>
+        /// Event fired each time the we add/remove securities from the data feed
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance that experienced the change in securities</param>
+        /// <param name="changes">The security additions and removals from the algorithm</param>
+        public override void OnSecuritiesChanged(QCAlgorithm algorithm, SecurityChanges changes)
+        {
+            // Get removed symbol and invalidate them in the insight collection
+            _removedSymbols = changes.RemovedSecurities.Select(x => x.Symbol).ToList();
```

Comment:
We could make `_removedSymbols` and array (`Symbol[]`) -- and then we won't need to use `ToArray()` below.I checked elsewhere in this file and it doesn't appear we're using any of the list features.